### PR TITLE
Software

### PR DESCRIPTION
**Pull request by rasmussaame**

### CODE REVIEW

File: software/python/basketBallDriver.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+        STATE.state_handler()
+        
+        if ord('q') == cv2.waitKey(1) & 0xFF:
+            break
+    ballThresholder.save()
+    poleThresholder.save()
+    yield
+
+
+def thresh(basket = "magenta"):
```
Comment (Akustav) [inline software/python/basketBallDriver.py:228]: Seems to be duplicate code. Decide what code you use and discard the rest

File: software/python/imageDetection.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    
+    def get_frames(self, aligned = True):
+        frames = self.pipeline.wait_for_frames()
+        frames = self.align.process(frames)
+        return np.asanyarray(frames.get_color_frame().get_data()), frames.get_depth_frame()
+
+
+class Thresholder:
+    def __init__(self, mode):
+        if mode.lower() not in ['rgb', 'hsv']:
```
Comment (Akustav) [inline software/python/imageDetection.py:80]: Enum can be used here to define values

**Review by Akustav — COMMENTED** [review]
Incremental update

File: software/python/robot.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    def spinRight(self):
+        speed = int(self.speed * 32767)
+        self.move(-speed, -speed, -speed)
+
+    
+    # speed1 - tagumine
+    # speed2 - parem
+    # speed3 - vasak
+    # motor range - 48-2047
+    def move(self, speed1 : int, speed2 : int, speed3 : int, throwerSpeed=0, disableFailsafe=0, delimiter=0xAAAA) -> None:
```
Comment (Akustav) [inline software/python/robot.py]: Currently can not see omnimotion, it seems to be differential drive with omniwheels

**Review by Akustav — COMMENTED** [review]
Interm. comments

File: software/python/state.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        return self.state
+
+    
+    def set_state(self, state):
+        self.state = state
+        self.state_start_time = time.time()
+
+
+    def state_go_to_opponent_basket(self):
+        if time.time() - self.state_start_time > STATE_OVERFLOW_TIME:
```
Comment (Akustav) [inline software/python/state.py:62]: Is this the only way to exit this state? What when you see a ball? Dot you want to chase it?

File: software/python/state.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+            self.state_start_time = time.time()
+        thresholded = self.create_thres_img_for_blob_detection(self.ball_thresholder)
+        cv2.imshow("Thresholded", thresholded)
+
+        keypoints = list(self.ball_detector.detect(thresholded))
+        keypoints.sort(key=lambda x: -x.size)  #  Biggest ball as first
+
+        self.frame_hsv = cv2.drawKeypoints(self.frame_hsv, keypoints, np.array([]), (0, 255, 0), cv2.DRAW_MATCHES_FLAGS_DRAW_RICH_KEYPOINTS)
+        if len(keypoints) == 0:
+            self.robot.set_speed(8 / 32767)
```
Comment (Akustav) [inline software/python/state.py:89]: Why this speed formula?

File: software/python/state.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+            self.robot.spin_left()
+        else:
+            kp = keypoints[0]
+            x_loc = kp.pt[0]
+            if kp.size < 45:
```
Comment (Akustav) [inline software/python/state.py:94]: Technically it is quite bad that you have image processing logic like pixel size filtering in the state logic. It should be handled in a base image processor with its set responsibilities

File: software/python/state.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        if time.time() - self.state_start_time > 5:
+            self.set_state(State.BALL)
+        thresholded = self.create_thres_img_for_blob_detection(self.pole_thresholder)
+        cv2.imshow("Pole thresholded", thresholded)
+        pole_keypoints = list(self.pole_detector.detect(thresholded))
+        self.frame_hsv = cv2.drawKeypoints(self.frame_hsv, pole_keypoints, np.array([]), (255, 0, 0), cv2.DRAW_MATCHES_FLAGS_DRAW_RICH_KEYPOINTS)
+        if len(pole_keypoints) == 0:
+            self.orbit_left(self.robot, 15, 0.57)
+        else:
+            pole_kp = pole_keypoints[0]
```
Comment (Akustav) [inline software/python/state.py:135]: Again image processing being done in the state machine

File: software/python/state.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+            elif x_loc > HALF_WIDTH + 8:
+                self.orbit_left(self.robot, 8, 0.57)
+            else:
+                self.robot.move(0, -20, 20, disable_failsafe=1)
+                self.delay_camera(0.25, self.cap)
+                self.go_forward(self.cap, self.robot, self.pole_thresholder, self.pole_detector)
+                self.set_state(State.BALL)
+
+
+    def create_thres_img_for_blob_detection(self, thresholder):
```
Comment (Akustav) [inline software/python/state.py:149]: This logic should be moved to a separate image processing module

File: software/python/robot.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        speed = int(self.speed * 32767)
+        self.move(-speed, -speed, -speed)
+
+    
+    # speed1 - tagumine
+    # speed2 - parem
+    # speed3 - vasak
+    # motor range - 48-2047
+    def move(self, speed1 : int, speed2 : int, speed3 : int, thrower_speed=0, disable_failsafe=0, delimiter=0xAAAA) -> None:
+        print(speed1, speed2, speed3, thrower_speed)
```
Comment (Akustav) [inline software/python/robot.py:71]: This method is heavily utilized to move but it just sends raw values. Not omni motion

File: software/python/state.py (hunk 5 above)
Comment (rasmussaame) [reply software/python/state.py:89]: Takes as parameter number from 0 to 1. Basically to easily compare the values to values that are sent to controller

File: software/python/state.py (hunk 4 above)
Comment (rasmussaame) [reply software/python/state.py:62]: Purpose of this task is to change the location and find new balls in this location, if there aren't any currently. It's good if we see ball at some point, but we will see some balls after moving as well so it doesn't matter so much

File: software/python/state.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+            if x_loc < (HALF_WIDTH - 15):
+                self.robot.spin_left(2)
+            elif x_loc >  (HALF_WIDTH + 15):
+                self.robot.spin_right(2)
+            else:
+                self.robot.stop()
+                quit(0)
+                start_time = time.time()
+                while time.time() - start_time < 1:
+                    self.robot.move(0, 20, -20)
```
Comment (Akustav) [inline software/python/state.py:89]: Why robot.move when you have omni_move

File: software/python/state.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        else:
+            pole_kp = pole_keypoints[0]
+            x_loc = pole_kp.pt[0]
+            loc = (int(x_loc), int(pole_kp.pt[1]))
+            if x_loc < HALF_WIDTH - 8:
+                self.orbit_right(self.robot, 8, 0.57)
+            elif x_loc > HALF_WIDTH + 8:
+                self.orbit_left(self.robot, 8, 0.57)
+            else:
+                self.robot.move(0, -20, 20, disable_failsafe=1)
```
Comment (Akustav) [inline software/python/state.py:112]: Why robot.move when you have omni_move

File: software/python/controller.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+                    data = conn.recv(1024)
+                    text = data.decode().rstrip()
+                    conn.send(b"ack")
+                    get_frames = False
+                    cv2.destroyAllWindows()
+                    try:
+                        obj = json.loads(text)
+                        if 'mode' in obj and 'speeds' in obj:
+                            if obj['mode'] == 'manual':
+                                robot.move(obj['speeds'][0], obj['speeds'][1], obj['speeds'][2], obj['speeds'][3])
```
Comment (Akustav) [inline software/python/controller.py:56]: Why robot.move when you have omni_move

File: software/python/basket_ball_driver.py [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+        frame, depth_frame = cap.get_frames()
+        frame_hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)
+        pole_thresholded = cv2.inRange(frame_hsv, pole_thresholder.get_low(), pole_thresholder.get_high())
+        pole_thresholded = 255 - pole_thresholded 
+        pole_thresholded = rect(pole_thresholded)
+        cv2.imshow("Pole thresholded", pole_thresholded)
+        pole_keypoints = list(pole_detector.detect(pole_thresholded))
+
+        if len(pole_keypoints) == 0:
+            robot.move(0, 20, -20, thr)
```
Comment (Akustav) [inline software/python/basket_ball_driver.py:181]: Why robot.move when you have omni_move

File: software/python/basket_ball_driver.py [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+        else:
+            x_loc = int(pole_keypoints[0].pt[0])
+            y_loc = int(pole_keypoints[0].pt[1])
+            y_coord = find_goal_location(pole_thresholded[1:,x_loc - 20 : x_loc + 20])
+            speed = int(get_thrower_speed_y_coord(y_coord / HEIGHT))
+            thr = speed
+            robot.move(0, 20, -20, speed)
```
Comment (Akustav) [inline software/python/basket_ball_driver.py:188]: Why robot.move when you have omni_move

File: software/python/basket_ball_driver.py [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+        frame, depth_frame = cap.get_frames()
+        frame_hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)
+        pole_thresholded = cv2.inRange(frame_hsv, pole_thresholder.get_low(), pole_thresholder.get_high())
+        pole_thresholded = 255 - pole_thresholded 
+        pole_thresholded = rect(pole_thresholded)
+        cv2.imshow("Pole thresholded", pole_thresholded)
+        pole_keypoints = list(pole_detector.detect(pole_thresholded))
+
+        if len(pole_keypoints) == 0:
+            robot.move(s1, s2, s3)
```
Comment (Akustav) [inline software/python/basket_ball_driver.py:150]: Why robot.move when you have omni_move

File: software/python/basket_ball_driver.py [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+
+def orbit_right(robot, max_val=35, a=0.75):
+    # NB TODO Vajab testimist ja kalibreerimist, proovi hiljem uuesti
+    # 35, -4, -4
+    s = [1, -0.5, -0.5]
+    or_s = [1, 1, 1]
+    
+    s1, s2, s3 = [((s[i] * a) + (or_s[i] * (1 - a)))for i in range(len(s))]
+    s1, s2, s3 = map_to_max(s1, s2, s3, max_val)
+    robot.move(s1, s2, s3)
```
Comment (Akustav) [inline software/python/basket_ball_driver.py:128]: Why robot.move when you have omni_move

File: software/python/basket_ball_driver.py [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+    # -35, 4, 4
+    
+    # liikumine vasakule
+    # NB TODO Vajab testimist ja kalibreerimist, proovi hiljem uuesti
+    s = [-1, 0.5, 0.5]
+    or_s = [-1, -1, -1]
+    
+    s1, s2, s3 = [((s[i] * a) + (or_s[i] * (1 - a)))for i in range(len(s))]
+    s1, s2, s3 = map_to_max(s1, s2, s3, max_val)
+    robot.move(s1, s2, s3)
```
Comment (Akustav) [inline software/python/basket_ball_driver.py:117]: Why robot.move when you have omni_move

File: software/python/robot.py [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+    
+    def move_left(self): # correct
+        self.move_omni(np.pi/2, 1000, 0, self.motors)
+    def move_right(self): # correct
+        self.move_omni(3*np.pi/2, 1000, 0, self.motors)    
+    def move_backward(self): # 
+        self.move_omni(np.pi, 1000, 0)
+    def move_forward(self): # correct
+        self.move_omni(0, 1000, 0)
+    def spin_right(self, speed=5): # correct
```
Comment (Akustav) [inline software/python/robot.py:41]: Why not have one method that takes either positive or negative values? Why have this level of abstraction?

**Review by Akustav — COMMENTED** [review]
Comment about omni

File: software/python/controller.py [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+                    conn.settimeout(0.01)
+                except socket.timeout:
+                    print("waiting")
+                    continue
+                break
+            print("Got connection")
+            generator = None
+            get_frames = False
+            while True:
+                try:
```
Comment (Akustav) [inline software/python/controller.py:67]: Any reconnect mechanism?

**Review by Akustav — APPROVED** [review]
Overall OK enough but the server reconnect mechanism

File: software/python/controller.py (hunk 19 above)
Comment (rasmussaame) [reply software/python/controller.py:67]: Yes, there's an outer while loop and another loop above these lines which handle connecting. If connection is lost, then the controller is waiting for somebody to connect with it. Achieved with timeouts and exceptions.

**Pull request merged**